test/xbmc.py

```diff
@@ -0,0 +1,155 @@
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2019, Dag Wieers (@dagwieers) <dag@wieers.com>
+# GNU General Public License v3.0 (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+''' This file implements the Kodi xbmc module, either using stubs or alternative functionality '''
+
+# pylint: disable=unused-argument
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import sys
+import os
+import json
+import time
+from xbmcextra import global_settings, import_language
+
+LOGDEBUG = 'Debug'
+LOGERROR = 'Error'
+LOGINFO = 'Info'
+LOGNOTICE = 'Notice'
+
+INFO_LABELS = {
+    'System.BuildVersion': '18.2',
+}
+
+REGIONS = {
+    'datelong': '%A, %e %B %Y',
+    'dateshort': '%Y-%m-%d',
+}
+
+GLOBAL_SETTINGS = global_settings()
+PO = import_language(language=GLOBAL_SETTINGS.get('locale.language'))
+
+
+class Keyboard:
+    ''' A stub implementation of the xbmc Keyboard class '''
+
+    def __init__(self, line='', heading=''):
+        ''' A stub constructor for the xbmc Keyboard class '''
+
+    def doModal(self, autoclose=0):
+        ''' A stub implementation for the xbmc Keyboard class doModal() method '''
+
+    def isConfirmed(self):
+        ''' A stub implementation for the xbmc Keyboard class isConfirmed() method '''
+        return True
+
+    def getText(self):
+        ''' A stub implementation for the xbmc Keyboard class getText() method '''
+        return 'unittest'
+
+
+class Monitor:
+    ''' A stub implementation of the xbmc Monitor class '''
+    def __init__(self, line='', heading=''):
+        ''' A stub constructor for the xbmc Monitor class '''
```

Comment:
Is it necessary for the correct functioning to insert stubs for kodi and inputstream modules?
what happens if kodi changes implementations?
Is it necessary to modify the work again? this seem too excessive?!

Reply:
Well, without this (and these are not just stubs, some of them actually perform functionality) you cannot run any tests that require them. So i you want to perform integration tests (e.g. run all menu items that users can run as well) you need this supporting code.

And it doesn't implement everything, but I did copy what we already implemented elsewhere, because it is more work to micro-manage this per project.

The Kodi python libraries are not usable and not available outside of Kodi (as it would require a complete Kodi to be available as well, which you cannot run in a container).

Reply:
The goal is to end up with something like this eventually:
https://travis-ci.org/pietje666/plugin.video.vrt.nu

A situation where you test every possible add-on call, and every possible function call, so you get a 100% test coverage of all you code (100% is theoretical because you cannot always test all failure cases). But 80% or 90% could be possible.

Reply:
now i understand, on these things you are much more informed than me..i am pleased
I would like to learn all this

Reply:
6 months ago I did not know any of this either, but I started to help with the VRT NU add-on and we developed this in order to ensure that every release was fully tested, and then we did the same last month for the InputStream Helper add-on which was having some problems.

If you have a high test coverage, you don't have to worry about a new release breaking something important, because every possible use case is covered by tests. And in case something is reported that wasn't covered, you can add more tests and improve your coverage. It was the only possible way to ensure that when we started add new functionality we could guarantee the existing functionality was working as before.

Reply:
This is the test coverage for VRT NU:
https://codecov.io/gh/pietje666/plugin.video.vrt.nu

It is not 100% because the content is geo-blocked and Travis runs from the US. So trying to authenticate and play content cannot be tested. However the menus and other functionality (like personal favorites, tv guide and search has test coverage of near 100%).
https://codecov.io/gh/pietje666/plugin.video.vrt.nu/tree/master/resources/lib